awd/article.py

```diff
@@ -18,57 +19,101 @@ class Article:
     metadata and binary content for uploading to our DSpace repository.
     """
 
-    def __init__(self, doi: str, metadata_url: str, content_url: str) -> None:
+    def __init__(
+        self,
+        doi: str,
+        metadata_url: str,
+        content_url: str,
+        doi_table: DoiProcessAttempt,
+        doi_table_items: list[DoiProcessAttempt],
+    ) -> None:
         """Initialize article instance.
 
         Args:
             doi: A digital object identifer (doi) for an article.
             metadata_url: The URL for retrieving metadata records.
-            content_url: The URL for retrieving article content
+            content_url: The URL for retrieving article content.
+            doi_table: The DOI table as a PynamoDB object.
+            doi_table_items: A list of the DOI table items as PynamoDB objects.
         """
         self.doi: str = doi
         self.metadata_url: str = metadata_url
         self.content_url: str = content_url
-        self.database_item = None
+        self.doi_table: DoiProcessAttempt = doi_table
+        self.doi_table_items: list[DoiProcessAttempt] = doi_table_items
         self.crossref_metadata: dict[str, Any]
         self.dspace_metadata: dict[str, Any]
         self.article_content: bytes
 
     def process(self) -> None:
+        """Run the complete article processing workflow."""
+        self.add_item_to_doi_table()
         self.get_and_validate_crossref_metadata()
         self.create_and_validate_dspace_metadata()
         self.get_and_validate_wiley_article_content()
 
-    def exists_in_database(self, database_items: list[dict[str, Any]]) -> bool:
-        """Validate that a DOI is NOT in the database and needs to be added.
-
-        Args:
-            database_items: A list of database items that may or may not contain the
-            specified DOI.
-        """
+    def exists_in_doi_table(self) -> bool:
+        """Validate that a DOI is NOT in the DOI table and needs to be added."""
         exists = False
-        if not any(doi_item["doi"] == self.doi for doi_item in database_items):
+        if any(doi_item.doi == self.doi for doi_item in self.doi_table_items):
             exists = True
-            logger.debug("%s added to database", self.doi)
+            logger.debug("%s added to DOI table", self.doi)
         return exists
```

Comment:
As noted below in `has_unprocessed_status` (commented there first), it might be possible to use the `DoiProcessAttempt` model more directly, particularly the `.get()` method, and avoid table scans and list comprehension that filter out results.

This one in particular would be a very nice method on the `DoiProcessAttempt` model.   In fact, I think it exists out of the box:

```python
# create instance of DoiProcessAttempt and supply the primary key (hash key)
doi_process_attempt = DoiProcessAttempt(doi="10.1002/kin.21154")

# utilize built-in exists method
doi_process_attempt.exists()
```